drivers/sx126x/sx126x_radio_hal.c

```diff
@@ -0,0 +1,668 @@
+    bool eagain = false;
+    sx126x_state_t state;
+    _get_state(dev, &state);
+
+    
+switch (op){
+    case IEEE802154_HAL_OP_TRANSMIT:
+       if (info) {
+            info->status = (dev->cad_detected) ? TX_STATUS_MEDIUM_BUSY : TX_STATUS_SUCCESS;
+        }
+
+        
+    break;
+
+    case IEEE802154_HAL_OP_SET_RX:
+        if(state!=STATE_RX) 
+            eagain = true;
+    break;
+
+    case IEEE802154_HAL_OP_SET_IDLE:
+        eagain = (state != STATE_IDLE);
+        break;
+
+    case IEEE802154_HAL_OP_CCA:
+            *((bool*) ctx) = !dev->cad_detected;
+    break;
+
+    default:
+        eagain = false;
+        break;
+    
+}
+
+    if (eagain) {
+       
+        return -EAGAIN;
+    }
+
+    return 0;
+}
+
+static int _len(ieee802154_dev_t *hal){
+    (void)hal;
+    sx126x_t *dev = hal->priv;
+    sx126x_rx_buffer_status_t rx_buffer_status;
+    sx126x_get_rx_buffer_status(dev, &rx_buffer_status);
+    return rx_buffer_status.pld_len_in_bytes;
+}
+
+static int _read(ieee802154_dev_t *hal, void *buf, size_t max_size, ieee802154_rx_info_t *info)
+{
+    (void)hal;
+    (void)buf;
+    (void)info;
+
+    sx126x_t* dev = hal->priv;
+        /* Getting information about last received packet */
+    
+    sx126x_rx_buffer_status_t rx_buffer_status;
+    sx126x_pkt_status_lora_t pkt_status;
+    sx126x_get_rx_buffer_status(dev, &rx_buffer_status);
+    sx126x_get_lora_pkt_status(dev, &pkt_status);
+    if (info) {
+        info->lqi = pkt_status.snr_pkt_in_db;
+        info->rssi = pkt_status.rssi_pkt_in_dbm;
+    }
+
+      /* Put PSDU to the output buffer */
+    
+    if (buf == NULL) {
+        return rx_buffer_status.pld_len_in_bytes;
+    }
+
+    if (rx_buffer_status.pld_len_in_bytes > max_size) {
+        return -ENOBUFS;
+    }
+
+    if (rx_buffer_status.pld_len_in_bytes < 3) {
+        return -EBADMSG;
+    }
+    sx126x_read_buffer(dev, rx_buffer_status.buffer_start_pointer, (uint8_t*)buf, rx_buffer_status.pld_len_in_bytes);
+    DEBUG("[sx126x] first 3 bytes of received packet: %d %d %d\n", *(uint8_t*)buf, *((uint8_t*)buf+1), *((uint8_t*)buf+2));
+    return rx_buffer_status.pld_len_in_bytes;
+}
+
+static int _set_cca_threshold(ieee802154_dev_t *hal, int8_t threshold)
+{
+    (void)hal;
+    (void)threshold;
+
+    return 0;
+}
+
+static int _config_phy(ieee802154_dev_t *hal, const ieee802154_phy_conf_t *conf){
+    (void)hal;
+    (void)conf;
+    sx126x_t *dev = hal->priv;
+    uint8_t channel = conf->channel;
+    int8_t pow = conf->pow;
+    if((channel > SX126X_CHAN_MAX))
+        return -EINVAL;
+    
+    sx126x_set_channel(dev, (channel)*SX126X_HAL_CHAN_SPACING + SX126X_HAL_CHAN_BASE);
+    if((pow < SX126X_POWER_MIN) && (pow > SX126X_POWER_MAX))
+        return -EINVAL;
+
+    sx126x_set_tx_params(dev, pow, SX126X_RAMP_10_US);
+    return 0;
+}
+
+static int _off(ieee802154_dev_t *hal)
+{
+    (void)hal;
+    sx126x_t *dev = hal->priv;
+    sx126x_set_sleep(dev, SX126X_SLEEP_CFG_COLD_START);
```

Comment:
```suggestion
    sx126x_set_sleep(dev, SX126X_SLEEP_CFG_WARM_START);
```

Cold start does not have RAM retention, which means the transceiver losses its state every time it sleeps. It is possible to implement a wake up procedure that recovers the radio, but it's probably easier to simply go to warm sleep.